devtools/gradle/src/main/java/io/quarkus/gradle/dependency/ConditionalDependenciesEnabler.java

```diff
@@ -0,0 +1,207 @@
+package io.quarkus.gradle.dependency;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.FileSystem;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Set;
+
+import org.gradle.api.GradleException;
+import org.gradle.api.Project;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.artifacts.Dependency;
+import org.gradle.api.artifacts.ModuleVersionIdentifier;
+import org.gradle.api.artifacts.ResolvedArtifact;
+import org.gradle.api.plugins.JavaPlugin;
+
+import io.quarkus.bootstrap.BootstrapConstants;
+import io.quarkus.bootstrap.util.ZipUtils;
+import io.quarkus.gradle.QuarkusPlugin;
+
+public class ConditionalDependenciesEnabler {
+
+    public static final String QUARKUS_EXTENSION_CONFIGURATION_NAME = "quarkusExtension";
+
+    private final Map<String, ExtensionDependency> featureVariants = new HashMap<>();
+    private final Project project;
+
+    public ConditionalDependenciesEnabler(Project project) {
+        this.project = project;
+    }
+
+    public void addConditionalDependencies() {
+
+        Configuration implementationCopy = DependencyUtils.duplicateConfiguration(project,
+                project.getConfigurations().getByName(JavaPlugin.IMPLEMENTATION_CONFIGURATION_NAME));
+
+        Set<ResolvedArtifact> runtimeArtifacts = implementationCopy.getResolvedConfiguration().getResolvedArtifacts();
+        List<ExtensionDependency> extensions = collectExtensionsForResolution(runtimeArtifacts);
+        featureVariants.putAll(extractFeatureVariants(extensions));
+
+        resolveConditionalDependencies(extensions, implementationCopy);
+
+        Configuration resolvedExtensions = DependencyUtils.duplicateConfiguration(project,
+                project.getConfigurations().getByName(JavaPlugin.IMPLEMENTATION_CONFIGURATION_NAME),
+                project.getConfigurations().getByName(QuarkusPlugin.DEV_MODE_CONFIGURATION_NAME));
```

Comment:
Do we always want the dev mode only dependencies in this configuration? Won't they leak into the application's lib dir?

Reply:
Could you clarify why this is needed including also the `exposeExtensionConfiguration`? I am probably missing something but it looks like the `configurationCopy` at this point is the proper runtime config. Is it not?

Reply:
I mean the `configurationCopy` before this line, right after `resolveConditionalDependencies`.

Reply:
This is needed by the model builder. This make sure all extensions are declared on a first level.

Reply:
The `configurationCopy` is correct but extensions are not first level module. We could keep it this way (without exposing the extension configuration) but we would need to update the builder to look for extensions in all the entire runtime classpath

Reply:
AFAIU, it's about collecting the conditional deployment deps. Otherwise, with the `configurationCopy` (which should probably be renamed to something more appropriate, btw) it would work like before.
Would you mind pushing it a little further? I was never happy with the way we resolve the deployment classpath and now with the conditional deps it looks like we have no choice but do it right.
Having identified all the runtime extension artifacts on the classpath (including the conditional ones), you should also know all their corresponding `deployment-artifact`s (they are in the same `extension-descriptor.properties` file).
Now, if you apply dependency substitution to the `configurationCopy` replacing all the runtime extension artifacts with the deployment ones, you'll get the proper application build classpath. But keep the `configurationCopy` as the source for application runtime `lib`. Does it make sense? Are you not exhausted yet? :)

Reply:
yes it totally make sense. I will do that. 

not exhausted yet :)